# Importing Libraries

In [53]:
# Important Libraries
import pandas as pd
import numpy as np
import os

# Importing Dataframes

In [54]:
# Telling python to remember a main folder path
path = r'/Users/datcaodo/Documents/22-03-2024 Instacart Basket Analysis'

In [56]:
# Importing dataset orders_wrangled.csv
df_ords_wrangled = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [57]:
# products csv
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [59]:
# The df.describe() function returns descriptive statistics for the numeric values in your dataframe.
df_ords_wrangled.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_last_order  
count        3.421083e+06       3.421083e+06           3.214874e+06  
mean         2.776219e+00       1.345202e+01           1.111484e+01  
std          2.046829e+00       4.226088e+00           9.206737e+00  
min          0.000000e+00       0.000000e+00           0.000000e+00  
25%          1.000000e+00       1.000000e+01           4.000000e+00  
50%          3.000000e+00       1.300000e+01           7.000000e+00  
75%          5.000000e+00       1.600000e+01           1.500000e+01  
max          6.000000e+00       2.300000e+01           3.000000e+01

In [60]:
# Create a dataframe
df_test = pd.DataFrame()

In [61]:
# Create a mixed type column 
df_test['mix'] = ['a','b',1, True]

#The second command, df_test['mix'] = ['a', 'b', 1, True], creates a new column, mix, 
#within df_test and fills it with numeric, string, and boolean values.

In [62]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [63]:
# Check for mixed types 

for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [64]:
# Changing mix to string because mostly names, otherwise numeric. 
df_test['mix'] = df_test['mix'].astype('str')

In [27]:
# from string to numeric would be:
# simply update the str within the astype() function to int64 or whichever numeric data type you want to use.

# Finding missing values

In [65]:
df_prods.isnull().sum()
#isnull() function is used to find missing observations, with “observations” here referring to entries in your dataframe. 
#Think of them like cells in Excel.

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

With a quick glance, you can see that the only column with missing values is the "product_name" column, and it’s missing 16 values.

In [66]:
# Creating a new dataframe, containing only those values within the 'product_name' column that meet the condtion isnull() = True.
df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [67]:
#Product_name with missing values
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

# Imputing missing values with the mean or median 

df['column with missings'].fillna(median value, inplace=True)

In [68]:
# The shape function returns the number of rows and columns in a dataframe
df_prods.shape

(49693, 5)

In [69]:
# Creating a new dataframe for a clean dataframe
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [70]:
# We can see that the 16 rows were filtered out.
df_prods_clean.shape

(49677, 5)

In [71]:
# Another way you can drop all missing values is via the following command:

df_prods.dropna(inplace = True)

#### If you wanted to use this command to drop only the NaNs from a particular column, the code would look like this:

df_prods.dropna(subset = [‘product_name’], inplace = True)

# Duplicates

In [72]:
#The following command will look for full duplicates within your dataframe:

df_dups = df_prods_clean[df_prods_clean.duplicated()]

#This code creates a new subset of df_prods_clean—df_dups—containing ONLY rows that are duplicates. 

In [73]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [74]:
df_prods_clean.shape

(49677, 5)

In [75]:
# New df that doesn't include the duplicates

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [76]:
# we can see that there are exactly 5 less rows
df_prods_clean_no_dups.shape

(49672, 5)

# Task 4.5 Questions

# Question 2

In [77]:
df_ords_wrangled.describe()

Unnamed: 0      order_id       user_id  order_number  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02   

       orders_day_of_week  order_hour_of_day  days_since_last_order  
count        3.421083e+06       3.421083e+06           3.214874e+06  
mean         2.776219e+00       1.345202e+01           1.111484e+01  
std          2.046829e+00       4.226088e+00           9.206737e+00  
min          0.000000e+00       0.000000e+00           0.000000e+00  
25%          1.000000e+00       1.000000e+01           4.000000e+00  
50%          3.000000e+00       1.300000e+01           7.000000e+00  
75%          5.000000e+00       1.600000e+01           1.500000e+01  
max          6.000000e+00       2.300000e+01           3.000000e+01

The minimum value of 0 in the order_dow column indicates that some orders were placed on day 0, which could be unusual or indicate a specific convention in the dataset.

Similarly, the minimum value of 0 in the order_hour_of_day column suggests that some orders were placed at hour 0 (midnight), which might need clarification or validation.

The maximum value of 30 in the days_since_prior_order column indicates that there are orders placed 30 days after the previous one, which may need verification as it could be an outlier or indicate a specific behavior in the dataset.

In [78]:
# Checking for Mixed-type data in df_ords dataframe

for col in df_ords_wrangled.columns.tolist():
  weird = (df_ords_wrangled[[col]].map(type) != df_ords_wrangled[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords_wrangled[weird]) > 0:
    print (col)

I haven't found any mixed data after running the code.

# Checking for missing values 

In [79]:
# Checking missing values in df_ords dataframe
df_ords_wrangled.isnull().sum()

Unnamed: 0                    0
order_id                      0
user_id                       0
order_number                  0
orders_day_of_week            0
order_hour_of_day             0
days_since_last_order    206209
dtype: int64

There are 206209 missing values in the 'days_since_prior_order' column

# Explanation
First-time Orders: For new users who are placing their first order, there would be no prior order to calculate the days since the previous order. Hence, the days_since_prior_order value would be missing.

Inactive Users: Users who haven't placed orders for an extended period might have missing values in this column because there are no recent orders to calculate the time since their last order.

Data Collection or Recording Issues: Missing values could also arise due to errors or issues in the data collection process, where the information about the days since the prior order was not recorded properly.

In [81]:
# Creating a new Dataframe containing the missing values
df_ords_nan = df_ords_wrangled[df_ords_wrangled['days_since_last_order'].isnull()==True]

In [82]:
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_last_order  
0                        8                    NaN  
11                      11                    NaN  
26                      14                    NaN  
39                      11                    NaN  
45                      12                    NaN  
...                    ...                    ...  
3420930                 12                    NaN  
3420934                 18                    NaN  
3421002                 19                    NaN  
3421019                 15                    NaN  
3421069                 11                    NaN  

[206209 rows x 7 columns]

In [83]:
df_ords_nan.shape

(206209, 7)

In [84]:
df_ords.describe()

order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

# Question 6 Checking for missing values and addressing the values
I would leave the missing values for now because there are 206209 rows that have missing values which is quiet a bit of data, if I would manipulate and delete them. 

What you can do here is to create a new boolean column, for example, "new_customer", which will be True if NaN and False if not. You do not have to change anything in the column days_since_prior_order, just create a new column and name it "new_customer" and fill it with TRUE and FALSE based on missing values.

In [85]:
# Address the missing values using an appropiate method
df_ords_clean = df_ords_wrangled
df_ords_clean['new_customer'] = df_ords['days_since_prior_order'].isnull()==True
df_ords_clean

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
1                 1   2398795        1             2                   3   
2                 2    473747        1             3                   3   
3                 3   2254736        1             4                   4   
4                 4    431534        1             5                   4   
...             ...       ...      ...           ...                 ...   
3421078     3421078   2266710   206209            10                   5   
3421079     3421079   1854736   206209            11                   4   
3421080     3421080    626363   206209            12                   1   
3421081     3421081   2977660   206209            13                   1   
3421082     3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_last_order  new_customer  
0                        8                    NaN          True  
1                        7                   15.0         False  
2                       12                   21.0         False  
3                        7                   29.0         False  
4                       15                   28.0         False  
...                    ...                    ...           ...  
3421078                 18                   29.0         False  
3421079                 10                   30.0         False  
3421080                 12                   18.0         False  
3421081                 12                    7.0         False  
3421082                 14                   30.0         False  

[3421083 rows x 8 columns]

In [86]:
df_ords_clean.shape

(3421083, 8)

# Question 7 and 8 Checking for Duplicates and Addressing the dataframe

In [87]:
#The following command will look for full duplicates within the dataframe:

df_ords_dups = df_ords_wrangled[df_ords_wrangled.duplicated()]

In [88]:
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_last_order, new_customer]
Index: []

In [89]:
df_ords_dups.shape

(0, 8)

There are no duplicated values in the dataframe of df_ords. Therefore, we don't need to address any values because there aren't any duplicate values.

# Exporting Data

In [91]:
df_ords_clean.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))

In [92]:
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))